src/CodeGen_Hexagon.cpp

```diff
@@ -175,6 +158,97 @@ Stmt sloppy_unpredicate_loads(Stmt s) {
     return SloppyUnpredicateLoads().mutate(s);
 }
 
+class InjectHVXLocks : public IRMutator2 {
+public:
+    InjectHVXLocks(const Target &t) : target(t) {
+        uses_hvx_var = Variable::make(Bool(), "uses_hvx");
+    }
+    bool uses_hvx = false;
+private:
+    Expr uses_hvx_var;
+    using IRMutator2::visit;
+    // Primarily, we do two things when we encounter a parallel for loop.
+    // First, we check if the paralell for loop uses_hvx and accordingly
+    // acqure_hvx_context i.e. acquire and release HVX locks.
+    // Then we insert a conditional unlock before the for loop, let's call
+    // this the prolog, and a conditional lock after the for loop which
+    // we shall call the epilog. So the code for a parallel loop that uses
+    // hvx should look like so.
+    //
+    // if (uses_hvx_var) {
+    //     halide_qurt_hvx_unlock();
+    // }
+    // parallel_for {
+    //     halide_qurt_hvx_lock();
+    //     ...
+    //     ...
+    //     halide_qurt_hvx_unlock();
+    // }
+    // if (uses_hvx_var) {
+    //     halide_qurt_hvx_lock();
+    // }
+    //
+    // When we move up to the enclosing scope we substitute the value of uses_hvx
+    // into the IR that should convert the conditionals to constants.
+    Stmt visit(const For *op) {
+        if (op->for_type == ForType::Parallel) {
+            bool old_uses_hvx = uses_hvx;
+            uses_hvx = false;
+
+            Stmt body = mutate(op->body);
+            Stmt s;
+            if (uses_hvx) {
+                body = acquire_hvx_context(body, target);
```

Comment:
I think you also need to substitute true for uses_hvx_var here. Basically, every time you do (or don't) acquire_hvx_context, you also need to substitute uses_hvx_var accordingly. So I would substitute true here, and substitute false in the else branch of this if.

We probably need a test for this. The test should have two nested parallel for loops, with vectorized code at each loop level.

Reply:
I did give this some thought last night and rejected the need to do what you say. Please correct me if I am wrong, but the only time we insert prolog & epilog is when the body of a parallel loop has a parallel for loop that uses_hvx (nested parallelism with vectorization). So, if uses_hvx is false, there should be no prologs and epilogs to substitute into.

Reply:
You insert a prolog and epilog around a parallel for loop with HVX code in the *body* of the loop (uses_hvx is true), but that says nothing about whether the code *outside* the loop uses HVX. You insert the prolog and epilog conditionalized on uses_hvx_var precisely because you can't know if uses_hvx will be true or not for that code later.

Later, when you actually decide the value of uses_hvx for that code, that's when you can replace uses_hvx_var with the real value.

Another way to look at it: you only have one uses_hvx_var that is shared amongst all instances in which the prolog and epilog are inserted. It's clearly not correct to substitute the same uses_hvx value for all of those.

Consider this example:

    parallel for y:
        parallel for x:
            // use HVX
        // use HVX

We need this to look like:

    parallel for y:
        qurt_hvx_lock();
        qurt_hvx_unlock();  // Stupid but intentional. We could try to optimize this case...
        parallel for x:
            qurt_hvx_lock();
            // use HVX
            qurt_hvx_unlock();
        qurt_hvx_lock();
        // use HVX
        qurt_hvx_unlock();

With your current code, you'll get this prior to the final IRMutator substitution:

    parallel for y:
        qurt_hvx_lock();
        if (uses_hvx_var):
            qurt_hvx_unlock();
        parallel for x:
            qurt_hvx_lock();
            // use HVX
            qurt_hvx_unlock();
        if (uses_hvx_var):
            qurt_hvx_lock();
        // use HVX
        qurt_hvx_unlock();

However, the final value of uses_hvx will be false, because it's outside the y loop and no code outside the y loop uses HVX. So, you'll substitute 'false' for uses_hvx_var, and get broken code:

    parallel for y:
        qurt_hvx_lock();
        parallel for x:
            qurt_hvx_lock();
            // use HVX
            qurt_hvx_unlock();
        // use HVX
        qurt_hvx_unlock();

We need a test that does this.